source/opt/optimizer.cpp

```diff
@@ -15,83 +15,107 @@
 #include "optimizer.hpp"
 
 #include "build_module.h"
+#include "make_unique.h"
 #include "pass_manager.h"
 #include "passes.h"
 
 namespace spvtools {
 
+struct Optimizer::PassToken::Impl {
+  Impl(std::unique_ptr<opt::Pass> p) : pass(std::move(p)) {}
+
+  std::unique_ptr<opt::Pass> pass;  // Internal implementation pass.
+};
+
+Optimizer::PassToken::PassToken(
+    std::unique_ptr<Optimizer::PassToken::Impl> impl)
+    : impl_(std::move(impl)) {}
+Optimizer::PassToken::PassToken(PassToken&& that)
+    : impl_(std::move(that.impl_)) {}
+
+Optimizer::PassToken& Optimizer::PassToken::operator=(PassToken&& that) {
+  impl_ = std::move(that.impl_);
+  return *this;
+}
+
+Optimizer::PassToken::~PassToken() {}
+
 struct Optimizer::Impl {
-  explicit Impl(spv_target_env env)
-      : target_env(env), message_consumer(IgnoreMessage), pass_manager() {}
+  explicit Impl(spv_target_env env) : target_env(env), pass_manager() {}
 
-  const spv_target_env target_env;
-  MessageConsumer message_consumer;
-  opt::PassManager pass_manager;
+  const spv_target_env target_env;  // Target environment.
+  opt::PassManager pass_manager;    // Internal implementation pass manager.
 };
 
 Optimizer::Optimizer(spv_target_env env) : impl_(new Impl(env)) {}
 
 Optimizer::~Optimizer() {}
 
 void Optimizer::SetMessageConsumer(MessageConsumer c) {
-  impl_->message_consumer = c;
+  // All passes' message consumer needs to be updated.
+  for (uint32_t i = 0; i < impl_->pass_manager.NumPasses(); ++i) {
```

Comment:
For another CL, but it seems we could simplify by adding begin() and end() to PassManager